### PR TITLE
Use Chip event inside emberAfEventControlSetDelayMS

### LINKED ISSUE #658: Provide access to CHIP event management system
#### Problem
How does CHIP manage system events? Does it require an RTOS? In Silabs ZCL we had a bare metal home grown Event system as well as RTOS support. We'll need something like this for ZCL in CHIP as well. 

 #### Proposed Solution
Provide access to Event management system

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem
Cluster events does not work currently.

 #### Summary of Changes
* Use Chip event for Cluster events

fixes #658

### CODE REVIEW

File: src/app/util/af-event.cpp [hunk 1]
```diff
@@ -65,7 +64,37 @@ struct EmberEventData
...
+// Stubs for Network related methods
+EmberEventControl emberAfPluginNetworkSteeringFinishSteeringEventControl;
+EmberEventControl emberAfPluginUpdateTcLinkKeyBeginTcLinkKeyUpdateEventControl;
+void emberAfPluginScanDispatchScanNetworkEventHandler(void){};
+void emberAfPluginNetworkSteeringFinishSteeringEventHandler(void){};
+void emberAfPluginUpdateTcLinkKeyBeginTcLinkKeyUpdateEventHandler(void){};
+// The Network Steering plugin defines a method that use emberEventControlSetInactive without
+// passing a pointer. But this method is unused, so let's override the method call for linking
+// to succeed.
+#define emberEventControlSetInactive(x) (void) 0
```
Comment (bzbarsky-apple) [inline src/app/util/af-event.cpp:86]: Could we just remove the unused `networkEventWrapper` bit from `af-gen-event.h`?  It'll have to be manually done when regenerating the gen files, but various other manual things have to happen there already, right?

File: src/app/util/af-event.cpp [hunk 2]
```diff
@@ -91,6 +120,23 @@ EmberEventData emAfEvents[] = {
     { NULL, NULL }
 };
 
+void EventControlHandler(chip::System::Layer * systemLayer, void * appState, chip::System::Error error)
+{
+    EmberEventControl * control = reinterpret_cast<EmberEventControl *>(appState);
+    if (control->status != EMBER_EVENT_INACTIVE)
+    {
+        for (size_t i = 0; i < sizeof(emAfEvents); i++)
```
Comment (bzbarsky-apple) [inline src/app/util/af-event.cpp:128]: This `sizeof()` makes no sense to me, unless each entry in `emAfEvents` is one-byte, which it's not.

How about doing it like so instead:
```
for (auto& event : emAfEvents) 
{
  if (event.control == control)
  {
    control->status = EMBER_EVENT_INACTIVE;
    event.handler();
    return;
  }
}
```
or so?

File: src/app/util/af-event.cpp [hunk 3]
```diff
@@ -124,7 +170,8 @@ EmberStatus emberAfEventControlSetDelayMS(EmberEventControl * control, uint32_t
 {
     if (delayMs <= EMBER_MAX_EVENT_CONTROL_DELAY_MS)
     {
-        // TODO: set a CHIP timer here emberEventControlSetDelayMS(*control, delayMs); #658
+        control->status = EMBER_EVENT_MS_TIME;
+        chip::DeviceLayer::SystemLayer.StartTimer(delayMs, EventControlHandler, control);
```
Comment (bzbarsky-apple) [inline src/app/util/af-event.cpp:174]: So `StartTimer` with a 0 delay automatically does the same thing as `ScheduleWork`?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/util/af-event.cpp (hunk 1 above)
Comment (vivien-apple) [reply src/app/util/af-event.cpp:86]: I actually plan to remove all that in a followup. I think we can just remove the Network steering plugin and all those will be removed.

File: src/app/util/af-event.cpp (hunk 2 above)
Comment (vivien-apple) [reply src/app/util/af-event.cpp:128]: Ah right. I always forgot C++11. Will do.

File: src/app/util/af-event.cpp (hunk 3 above)
Comment (vivien-apple) [reply src/app/util/af-event.cpp:174]: I can not really confirm that. I believe this is platform dependent.

File: src/app/util/af-event.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/util/af-event.cpp:86]: Works for me.  Just need to make sure the followup happens.

File: src/app/util/af-event.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/util/af-event.cpp:174]: OK.  The original impl of `emberAfEventControlSetDelayMS` looks like this:
```
EmberStatus emberAfEventControlSetDelayMS(EmberEventControl *control,
                                          uint32_t delayMs)
{
  if (delayMs == 0) {
    emberEventControlSetActive(*control);
  } else if (delayMs <= EMBER_MAX_EVENT_CONTROL_DELAY_MS) {
    emberEventControlSetDelayMS(*control, delayMs);
```
so if we want to preserve those semantics and don't know whether `StartTimer` guarantees them, I think we should introduce that check-for-0 thing.

Comment (rwalker-apple) [reply src/app/util/af-event.cpp:174]: I can confirm that starttimer() of zero runs the timer immediately.  Also: SetActive() calling the handler immediately can cause a stack overflow (i.e. where running the handler merely reschedules the handler with 0).

Comment (rwalker-apple) [reply src/app/util/af-event.cpp:174]: schedulework() looks exactly like a zero-based timer without the reverse sort.  it's not really platform-dependent, as all the lists are in the CHIP system layer.

**Review by rwalker-apple — APPROVED** [review]
lGtm

File: src/app/util/af-event.cpp (hunk 1 above)
Comment (vivien-apple) [reply src/app/util/af-event.cpp:86]: Got it. Let me remove it first then. See #2923

Comment (vivien-apple) [reply src/app/util/af-event.cpp:86]: >  It'll have to be manually done when regenerating the gen files, but various other manual things have to happen there already, right?

AFAIK the 4 manual things that needs to be done for those generated file are:
 * Add back the License
 * Run clang-format
 * Update the EMBER_AF_GENERATED_REPORTING_CONFIG_DEFAULTS macro from gen/endpoint_config.h
 * Add #include "debug-printing-test.h" to gen/gen_config.h

Those are already a bit painful. I wish we won't add more !

Comment (vivien-apple) [reply src/app/util/af-event.cpp:86]: Sigh. It seems like that even if #2923 let us remove some stubs from the PR, `clusterTickWrapper` from `gen/af-gen-event.h` still calls `emberEventControlSetInactive` as well as `emberAfPushEndpointNetworkIndex` and `emberAfPopNetworkIndex`

So I will add a new manual step to the list as you suggested. I still think it worth landing #2923 though.

**Review by mspang — APPROVED** [review]

File: src/app/util/af-event.cpp (hunk 2 above)
Comment (vivien-apple) [reply src/app/util/af-event.cpp]: Done.

File: examples/wifi-echo/server/esp32/main/gen/af-gen-event.h [hunk 4]
```diff
@@ -47,10 +47,8 @@
     extern void emberAfPluginReportingTickEventHandler(void);                                                                      \
     static void clusterTickWrapper(EmberEventControl * control, EmberAfTickFunction callback, uint8_t endpoint)                    \
     {                                                                                                                              \
-        emberAfPushEndpointNetworkIndex(endpoint);                                                                                 \
```
Comment (bzbarsky-apple) [inline examples/wifi-echo/server/esp32/main/gen/af-gen-event.h:50]: Could you please comment those network index bits out, not remove them, so we can decide what to do with them when we deal with the relevant issue?

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@saurabhst @jelderton @hawk248 @BroderickCarlin ?

**Review by BroderickCarlin — APPROVED** [review]

**Pull request merged**